src/mito2/src/compaction/twcs.rs

```diff
@@ -195,24 +201,88 @@ impl Picker for TwcsPicker {
     }
 }
 
+#[derive(Default)]
+struct Window {
+    start: Option<Timestamp>,
+    end: Option<Timestamp>,
+    files: Vec<FileHandle>,
+    time_window: i64,
+    overlapping: bool,
+}
+
+impl Window {
+    /// # Panics
+    /// If `files` is empty.
+    fn range(&self) -> (Timestamp, Timestamp) {
+        assert!(!self.files.is_empty());
+        (self.start.unwrap(), self.end.unwrap())
+    }
+
+    /// Adds a new file to window and updates time range.
+    fn add_file(&mut self, file: FileHandle) {
+        let (start, end) = file.time_range();
+        self.start = Some(self.start.map_or(start, |e| e.min(start)));
+        self.end = Some(self.end.map_or(end, |e| e.max(end)));
+        self.files.push(file);
+    }
+}
+
 /// Assigns files to windows with predefined window size (in seconds) by their max timestamps.
 fn assign_to_windows<'a>(
     files: impl Iterator<Item = &'a FileHandle>,
     time_window_size: i64,
-) -> BTreeMap<i64, Vec<FileHandle>> {
-    let mut windows: BTreeMap<i64, Vec<FileHandle>> = BTreeMap::new();
+) -> BTreeMap<i64, Window> {
+    let mut windows: HashMap<i64, Window> = HashMap::new();
     // Iterates all files and assign to time windows according to max timestamp
-    for file in files {
-        let (_, end) = file.time_range();
+    for f in files {
+        let (_, end) = f.time_range();
         let time_window = end
             .convert_to(TimeUnit::Second)
             .unwrap()
             .value()
             .align_to_ceil_by_bucket(time_window_size)
             .unwrap_or(i64::MIN);
-        windows.entry(time_window).or_default().push(file.clone());
+        let window = windows.entry(time_window).or_default();
+        window.time_window = time_window;
+        window.add_file(f.clone());
     }
-    windows
+    if windows.is_empty() {
+        return BTreeMap::new();
+    }
+
+    let mut windows = windows.into_values().collect::<Vec<_>>();
+    windows.sort_unstable_by(|l, r| l.start.cmp(&r.start).then(l.end.cmp(&r.end).reverse()));
+
+    let mut current_range: Option<(Timestamp, Timestamp)> =
+        Some((windows[0].start.unwrap(), windows[0].end.unwrap()));
+
+    for idx in 1..windows.len() {
+        match current_range {
+            None => {
+                current_range = Some(windows[idx - 1].range());
+                continue;
+            }
+            Some(current) => {
+                let next_range = windows[idx].range();
+                if overlaps(&current, &next_range) {
+                    windows[idx - 1].overlapping = true;
+                    windows[idx].overlapping = true;
+                }
+                current_range = Some((current.0.min(next_range.0), current.1.max(next_range.1)));
```

Comment:
I guess we should set the range to None or the next range if `overlaps()` returns false. Do unit tests cover this?